src/lightning/pytorch/utilities/restricted_classmethod.py

```diff
@@ -0,0 +1,34 @@
+"""
+Implements a decorator @_restricted_classmethod that is a drop-in replacement for
+@classmethod, but raises an exception when the decorated method is called on an instance
+instead of a class type.
+"""
+from types import MethodType
+from typing import Callable, Concatenate, Generic, ParamSpec, TYPE_CHECKING, TypeVar
+
+_T = TypeVar("_T")  # type of the method owner
+_P = ParamSpec("_P")  # parameters of the decorated method
+_R_co = TypeVar("_R_co", covariant=True)  # return type of the decorated method
+
+
+class _restricted_classmethod_impl(Generic[_T, _P, _R_co]):
+    """
+    Custom `classmethod` that raises an exception when the classmethod is
+    called on an instance and not the class type.
+    """
+
+    def __init__(self, method: Callable[Concatenate[_T, _P], _R_co]) -> None:
+        self.method = method
+
+    def __get__(self, instance: _T | None, cls: type[_T]) -> Callable[_P, _R_co]:
+        if instance is not None:
+            raise TypeError(
+                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. "
+                f"Please call it on the class type and make sure the return value is used."
+            )
```

Comment:
Hey @carmocca @kwsp I think this is an important PR because this mistake happens very frequently. I would suggest we put the jit scripting at a lower priority and compromise on making a hack to enable this PR. I played around with the following check and it seems to work where `torch.is_jit_scripting()` fails. 

```suggestion
        if instance is not None and not any("torch/jit" in s.filename for s in inspect.stack()):
            raise TypeError(
                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance. "
                f"Please call it on the class type and make sure the return value is used."
            )
```

Should we try this?

Reply:
Fine with me if it works. We should evaluate removing support for `torch.jit` soon.